source/agora/script/Engine.d

```diff
@@ -0,0 +1,1213 @@
+
+        If the lock is a `Lock.Key` type, the unlock must only
+        contain a `signature`.
+        If the lock is a `Lock.KeyHash` type, the unlock must contain a
+        `signature, key` tuple.
+
+        Params:
+            lock = must contain a `pubkey` or a `hash`
+            unlock = must contain a `signature` or `signature, key` tuple
+
+        Returns:
+            null if there were no errors,
+            or a string explaining the reason execution failed
+
+    ***************************************************************************/
+
+    private string handleBasicPayment (in Lock lock, in Unlock unlock,
+        in Transaction tx) nothrow @safe
+    {
+        // assumed sizes
+        static assert(Point.sizeof == 32);
+        static assert(Hash.sizeof == 64);
+
+        switch (lock.type)
+        {
+        case LockType.Key:
+            if (lock.bytes.length != Point.sizeof)
+                return "LockType.Key requires 32-byte key argument in the lock script";
+            if (!isValidPointBytes(lock.bytes))
+                return "LockType.Key 32-byte public key in lock script is invalid";
+            const Point key = Point(lock.bytes);
+
+            if (unlock.bytes.length != Signature.sizeof)
+                return "LockType.Key requires a 64-byte signature in the unlock script";
+            const sig = Signature(unlock.bytes);
+            if (!Schnorr.verify(key, sig, tx))
+                return "LockType.Key signature in unlock script failed validation";
+
+            break;
+
+        case LockType.KeyHash:
+            if (lock.bytes.length != Hash.sizeof)
+                return "LockType.KeyHash requires a 64-byte key hash argument in the lock script";
+            const Hash key_hash = Hash(lock.bytes);
+
+            const(ubyte)[] bytes = unlock.bytes;
+            if (bytes.length != Signature.sizeof + Point.sizeof)
+                return "LockType.KeyHash requires a 64-byte signature "
+                     ~ "and a 32-byte key in the unlock script";
+            const sig = Signature(bytes[0 .. Signature.sizeof]);
+            bytes.popFrontN(Signature.sizeof);
+
+            if (!isValidPointBytes(bytes))
+                return "LockType.KeyHash public key in unlock script is invalid";
+            const Point key = Point(bytes);
+
+            if (!Schnorr.verify(key, sig, tx))
+                return "LockType.KeyHash signature in unlock script failed validation";
+
+            break;
+
+        default:
+            assert(0);
+        }
+
+        return null;
+    }
+
+    /***************************************************************************
+
+        Execute a `LockType.Script` type of lock script with the associated
+        unlock script.
+
+        The unlock script may only contain stack pushes.
+        The unlock script is ran, producing a stack.
+
+        Thereafter, the lock script will run with the stack of the
+        unlock script.
+
+        Params:
```

Comment:
Add a note why the two scripts should not be concatenated together before executio . Someone might think they're smart and try to optimize this, but it would cause a security vulnerability.